### PR TITLE
Add empty! to simple interface

### PR DESCRIPTION
**Pull request by ChrisRackauckas**
Takes note of https://github.com/JuliaDiffEq/Sundials.jl/pull/112#issuecomment-289403788 and adds a manual free in the simplified interface.

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
@alyst @tkelman 

Not a full solution, but does one interface.

**Comment by tkelman** [comment]
what about in the two cases that used to use `finally` in `src/common.jl` ?

**Comment by ChrisRackauckas** [comment]
That's why I say this is for the simplified interface, and it's a partial solution. We know that for `simple.jl`, this is the right way to do it, so I wanted to get this out of the way. For the interface of `common.jl`, that one will need something more complex (https://github.com/JuliaDiffEq/Sundials.jl/pull/112#issuecomment-289845251) or it will block the next developments, so it can't take this same approach.

**Comment by ChrisRackauckas** [comment]
I added the safety measures for the common interface for now. In the future this will change, but this issue with custom allocators shouldn't block the v0.6-compatibility release.

**Comment by alyst** [comment]
Are there any existing use-cases that would need active GC of Sundial objects?

Once I was using Sundials.jl/CVODE inside a function that was then optimized using evolutionary programming.
The ODE system was small, tens of variables and tens of timepoints, nevertheless the memory usage was stable across millions of iterations.
In that case the RHS of ODE didn't change, so I was using `CVodeReInit()`.

File: src/common.jl [hunk 1]
```diff
@@ -60,8 +60,9 @@ function solve{uType,tType,isinplace,F,Method,LinearSolver}(
     end
 
     mem_ptr = CVodeCreate(alg_code, method_code)
-    (mem_ptr == C_NULL) && error("Failed to allocate CVODE solver object")
```
Comment (alyst) [inline src/common.jl]: Why do you need to initialize `mem` if `mem_ptr == C_NULL`?

File: src/handle.jl [hunk 2]
```diff
@@ -29,25 +29,33 @@ typealias KINMemPtr Ptr{KINMem}
    Manages automatic destruction of the referenced objects when it is
    no longer in use.
 """
-immutable Handle{T <: AbstractSundialsObject}
+immutable Handle{T}
```
Comment (alyst) [inline src/handle.jl]: I think we don't need to allow any `T`

**Comment by tkelman** [comment]
`CVodeFree` etc were there for a reason I'm assuming? Not super familiar with how much allocation goes on inside the library though.

**Comment by ChrisRackauckas** [comment]
>Are there any existing use-cases that would need active GC of Sundial objects?

Not yet, because the [integrator interface](http://docs.juliadiffeq.org/latest/basics/integrator.html) for [event-handling](http://docs.juliadiffeq.org/latest/features/callback_functions.html) hasn't been put on it yet. But that's a bad way to think about it, because those were being blocked from being implemented by the fact that it all needed to sit in a try/finally block, which makes it impossible to do the dynamic/iterative approach.

A case where GC is needed is when it's used for things like parameter estimation. That will create lots of separate independent (and sometimes parallel) solutions/solvers. In the future, that probably should use `CVodeReInit`, and all of the other solvers should use pre-allocated versions using the solution of a previous solution, but for now this kind of functionality has to be built agnostic to pre-allocation until we have a good dispatch solution for that.

Also, there were tales that the old approach wasn't working:

https://github.com/JuliaDiffEq/Sundials.jl/issues/111

I could never confirm that though.

File: src/handle.jl [hunk 3]
```diff
@@ -29,25 +29,33 @@ typealias KINMemPtr Ptr{KINMem}
...
     ptr_ref::Ref{Ptr{T}} # pointer to a pointer
 
-    @compat function (::Type{Handle}){T}(ptr::Ptr{T})
+    @compat function (::Type{Handle}){T<: AbstractSundialsObject}(ptr::Ptr{T})
         (ptr == C_NULL) && throw(ArgumentError("Null pointer passed to Handle()"))
         h = new{T}(Ref{Ptr{T}}(ptr))
         finalizer(h.ptr_ref, release_handle)
         return h
     end
+    @compat function (::Type{Handle})(ptr::Ptr{Void})
```
Comment (alyst) [inline src/handle.jl]: It's not necessary, because if e.g. `CVodeCreate()` returns null, it would be of type `CVODEMemPtr===Ptr{CVODEMem}`. Allowing `Void` or any other non `AbstractSundialsObject` types would just hide potential misuses of the `Handle`. Also, it makes Sundials type-unstable with a big impact on performance.

File: src/handle.jl [hunk 4]
```diff
@@ -29,25 +29,33 @@ typealias KINMemPtr Ptr{KINMem}
...
+        finalizer(h.ptr_ref, release_handle)
+        return h
+    end
 end
 
 Base.convert{T}(::Type{Ptr{T}}, h::Handle{T}) = h.ptr_ref[]
 Base.convert{T}(::Type{Ptr{Ptr{T}}}, h::Handle{T}) = convert(Ptr{Ptr{T}}, h.ptr_ref[])
 
 release_handle{T}(ptr_ref::Ref{Ptr{T}}) = throw(MethodError("Freeing objects of type $T not supported"))
+release_handle(ptr_ref::Ref{Ptr{Void}}) = nothing
```
Comment (alyst) [inline src/handle.jl]: I think we just need to introduce the `C_NULL` check in each `release_handle()`.

File: src/handle.jl [hunk 5]
```diff
@@ -29,25 +29,33 @@ typealias KINMemPtr Ptr{KINMem}
...
 release_handle(ptr_ref::Ref{Ptr{KINMem}}) = KINFree(ptr_ref)
 release_handle(ptr_ref::Ref{Ptr{CVODEMem}}) = CVodeFree(ptr_ref)
 release_handle(ptr_ref::Ref{Ptr{IDAMem}}) = IDAFree(ptr_ref)
 
+Base.empty!{T}(h::Handle{T}) = release_handle(h.ptr_ref)
```
Comment (alyst) [inline src/handle.jl]: Also it makes sense to define `Base.isempty(Handle)` method

**Comment by tkelman** [comment]
Looks like a lot of the early issues here were reporting memory leaks of various kinds, so might be good to distill a few of those into test / benchmark cases if the existing tests wouldn't flag unusually high memory consumption in any way.

**Comment by alyst** [comment]
@ChrisRackauckas Do you know whether Sundials developers already considered custom memory allocators? If not yet, maybe it makes sense to ask what is their opinion? (Although I don't understand how it would be handled by Julia, because not every object that Sundials allocates would be exposed to Julia).

File: src/common.jl [hunk 6]
```diff
@@ -60,7 +60,7 @@ function solve{uType,tType,isinplace,F,Method,LinearSolver}(
...
+    (mem_ptr == C_NULL) && (release_handle!(mem_ptr);error("Failed to allocate CVODE solver object"))
```
Comment (alyst) [inline src/common.jl]: If `mem_ptr == C_NULL`, then nothing was allocated, no need to release.

Comment (ChrisRackauckas) [inline src/common.jl]: oh yes

**Comment by ChrisRackauckas** [comment]
>@ChrisRackauckas Do you know whether Sundials developers already considered custom memory allocators? If not yet, maybe it makes sense to ask what is their opinion? (Although I don't understand how it would be handled by Julia, because not every object that Sundials allocates would be exposed to Julia).

My C-foo is very lacking (if you couldn't tell), but from what I can tell from my internet reading since @tkelman first introduced me to that custom allocators exist, Google searches turn up empty. We can probably ask on their forum.

**Comment by ChrisRackauckas** [comment]
>Looks like a lot of the early issues here were reporting memory leaks of various kinds, so might be good to distill a few of those into test / benchmark cases if the existing tests wouldn't flag unusually high memory consumption in any way.

In all of the latest ones, I couldn't confirm it on either Windows or Linux, and no one had an MWE. Some of the issues I only knew about from reading comments in other peoples codes! I have tried a few different things which would allocate a lot, but it always seemed to GC correctly for me... so I am kind of going blind there.

I'm thinking a good direction to go then is to essentially augment what we had before (manual freeing) with finalizers, and hopefully that helps with whatever issues were around. That would be fine for a v0.6-compatibility release (which actually only needs https://github.com/JuliaDiffEq/Sundials.jl/commit/1f052b10fa349343692f23cc246bbadd363a10ff , but it would be nice to also include some attempt at the reported memory problems) to just do this, and leave event handling and all of that for another day. 

And if getting the GC to work in a way that's good for dynamic controls is too difficult, that's okay. I see Sundials' role as a stable stop-gap until we have more native Julia methods. My roadmap is to try and finish (1) the version upgrade, (2) expose ARKODE (with a simplified interface), (3) put as much of the diffeq interface on it as possible, and (4) add some event-handling, then only do maintenance. Since my C++ skills are clearly lacking (thanks for all of the help BTW!), I'm just trying to pick the low hanging fruit out of those 4 to reasonably complete the package, while most of my attention is spent building an alternative.

**Comment by ChrisRackauckas** [comment]
I resolved conflicts with master. This version of the PR calls `empty!` at the end of any simplified or common interface routine. Is this good @alyst ?

File: src/handle.jl [hunk 7]
```diff
@@ -32,7 +32,7 @@ typealias KINMemPtr Ptr{KINMem}
 immutable Handle{T <: AbstractSundialsObject}
     ptr_ref::Ref{Ptr{T}} # pointer to a pointer
 
-    @compat function (::Type{Handle}){T}(ptr::Ptr{T})
+    @compat function (::Type{Handle}){T <: AbstractSundialsObject}(ptr::Ptr{T})
         (ptr == C_NULL) && throw(ArgumentError("Null pointer passed to Handle()"))
```
Comment (alyst) [inline src/handle.jl]: Since `Handle` can handle nulls, I would allow nulls in the ctor as well.

**Comment by alyst** [comment]
@ChrisRackauckas Looks good. I would allow constructing handles pointing to NULL.
And I think you have to add the tests for the `Handle`, because ATM the Sundials.jl tests are very superficial and Handle is quite important functionality to have dedicated tests.
See my comments above regarding the tests that could be done.

**Comment by ChrisRackauckas** [comment]
Adding the requested tests, found some errors, smashed them.

File: test/runtests.jl [hunk 8]
```diff
@@ -61,6 +61,11 @@ include(joinpath(examples_path, "kinsol_mkinTest.jl"))
 @test abs(minimum(residual)) < 1e-5
 end
 
+println("== handle tests")
+let
+@testset "Handle Tests" begin include("handle_tests.jl") end
```
Comment (alyst) [inline test/runtests.jl]: I think here `@testset` is not necessary, but printing the file being executed would be nice.
I don't know what is the most "idiomatic" way to organize Julia package tests, but you can look into e.g. [DataTables.jl runtests.jl](https://github.com/JuliaData/DataTables.jl/blob/master/test/runtests.jl) for how it is done (a tiny infrastructure to run tests from multiple files).

File: test/handle_tests.jl [hunk 9]
```diff
@@ -0,0 +1,15 @@
+println("Test empty isempty")
```
Comment (alyst) [inline test/handle_tests.jl]: This is a candidate for `@testset`

File: test/handle_tests.jl [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+@test !isempty(h1)
+@test !isempty(h2)
+
+empty!(h1)
+
+@test isempty(h1)
+@test isempty(h2)
+
+println("Handle pointing to NULL")
```
Comment (alyst) [inline test/handle_tests.jl]: This is as well `@testset` candidate. I would avoid to much explicit printing in tests. If the test fails, then `@testset` macro is supposed to show what and where have failed. If the tests are fine, this line doesn't provide useful info.

File: test/handle_tests.jl (hunk 9 above)
Comment (ChrisRackauckas) [inline test/handle_tests.jl]: This can't be a `@testset` because I use the emptied `h1` in the next part.

File: test/runtests.jl (hunk 8 above)
Comment (ChrisRackauckas) [inline test/runtests.jl]: Instead of printing the file, if you just put it in a testset you get the same information. All of the Sundials tests should be testsets instead of let blocks. I only did the let block thing to match the current style, but they should really match the more modern style of diffeq:

https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/master/test/runtests.jl

The DataTables.jl example is complicated in unnecessary ways. I would like to avoid that style.

Comment (alyst) [inline test/runtests.jl]: Ok, I see, looks fine.

**Comment by ChrisRackauckas** [comment]
>because ATM the Sundials.jl tests are very superficial 

I agree that the Sundials.jl tests are extremely bad. Mostly lacking (the test coverage is an abysmal <20%!), but also using old-school `let` blocks in unnecessary ways. I hope to get around to making that its own PR though, since there's quite a lot to do there.

File: test/runtests.jl (hunk 8 above)
Comment (alyst) [inline test/runtests.jl]: I would then remove the `println()` or think about avoiding typing the name of the test set in 3 places as it tends to be desynchronized.

Comment (alyst) [inline test/runtests.jl]: But maybe this polishing could be done in separate PR.

Comment (ChrisRackauckas) [inline test/runtests.jl]: I think it's a separate PR, because we should really want to cleanup those let blocks at the same time, actually test values on the other examples, stop printing so many times in the common tests, etc. There's a lot to do here, so for now I semi-matched the old style.

File: test/handle_tests.jl [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+h2 = h1
+
+@test !isempty(h1)
+@test !isempty(h2)
+
+empty!(h1)
+
+@test isempty(h1)
+@test isempty(h2)
+
```
Comment (alyst) [inline test/handle_tests.jl]: Could you also add a test that second `empty!(h1)` would not throw?
